### ISSUE TITLE
Fatal compiler error with Variant

### ISSUE THREAD

**Issue by D-a-n-i-l-o** [blitz-research/monkey2#482]
```monkey
#Import "<monkey>"

Using monkey.types

Function Main()

    Print Variant( 1.23 ).Type.Name
    Print Variant( 1234 ).Type.Name
    Print Variant( True ).Type.Name
    Print Variant( "Hi" ).Type.Name

    Print "-----------------------"

    Print Variant( Bool(1)                      ).Type.Name
    Print Variant( Byte(-20)                    ).Type.Name
    Print Variant( UByte($FF)                   ).Type.Name
    Print Variant( Short( -32768)               ).Type.Name
    Print Variant( UShort( $FFFF)               ).Type.Name
    'Print Variant( Int( -2147483648 )           ).Type.Name
    Print Variant( UInt( $FFFFFFFF )            ).Type.Name
    'Print Variant( Long( -9223372036854775808 ) ).Type.Name
    'Print Variant( ULong( $FFFFFFFFFFFFFFFF )   ).Type.Name
    Print Variant( Float( 123.456 )             ).Type.Name
    Print Variant( Double( 789.123 )            ).Type.Name
    Print Variant( String( 12 )                 ).Type.Name
    Print Variant( New Object                   ).Type.Name

    Print "-----------------------"
    
    Print Variant( New Int[10] ).Type.Name

End
```
With the 3 commented types: Int, Long, and ULong I get a compiler error:
```
***** Fatal mx2cc error *****

Internal mx2cc build error
```

It is a problem with the C++ code.

Tested on macOS.

Error output:
```
untitled2.buildv1.1.15/macos_release_mx/src/untitled2_untitled2.cpp:28:12: note: in instantiation of function template specialization 'bbVariant::bbVariant<unsigned short>' requested here
  bb_print(bbVariant(bbUShort(65535)).getType()->getName());
           ^
In file included from /Users/name/monkey2/tmp/untitled2.buildv1.1.15/macos_release_mx/src/untitled2_untitled2.cpp:2:
In file included from /Users/name/monkey2/tmp/untitled2.buildv1.1.15/macos_release_mx/include/untitled2_untitled2.h:5:
In file included from /Users/name/monkey2/modules/monkey/native/bbmonkey.h:14:
In file included from /Users/name/monkey2/modules/monkey/native/bbvariant.h:5:
/Users/name/monkey2/modules/monkey/native/bbtypeinfo.h:85:9: error: call to 'bbGetType' is ambiguous
        return bbGetType( *(T*)0 );
               ^~~~~~~~~
```

**Comment by seyhajin** [comment]
I have tested your code and i don't have any build error except an typeinfo 'unknown' for 'Int' : 
```
Float
Int
Bool
String
-----------------------
Bool
Byte
UByte
Short
UShort
Unknown@140702302300064 '<==========
UInt
Long
ULong
Float
Double
String
Object
-----------------------
Int[]
```

But if i change 'Int' value from '-2147483648' to '-2147483647', I have no more mistakes 👍 

```
Float
Int
Bool
String
-----------------------
Bool
Byte
UByte
Short
UShort
Int '<=============
UInt
Long
ULong
Float
Double
String
Object
-----------------------
Int[]
```

For information my 'mx2cc_windows' has been recompiled with MSVC 2017 64 bits.

**Comment by D-a-n-i-l-o** [comment]
Okay, seems to be a macOS-only issue then?

I am using:
* macOS Mojave, Version 10.14.5 (18F132)
* Xcode Version 10.2.1 (10E1001)

With the reduced code:
```monkey
#Import "<monkey>"

Using monkey.types

Function Main()
    Print Variant( Int( -2147483647 ) ).Type.Name ' works
    Print Variant( Long( 1 )          ).Type.Name ' error on macOS
End
```
The error message is:
```

Mx2cc version 1.1.15

***** Making app '/Users/danilo/Monkey/monkey2/tmp/untitled3.monkey2' (macos release x64 gcc mx) *****

Parsing...
Semanting...
Translating...
Compiling...
Build error: System command failed:

g++ -c -std=c++14 -mmacosx-version-min=10.9 -Wno-deprecated-declarations -Wno-tautological-pointer-compare -Wno-undefined-bool-conversion -Wno-int-to-void-pointer-cast -Wno-inconsistent-missing-override -Wno-logical-op-parentheses -Wno-parentheses-equality -O3 -DNDEBUG -I"/Users/danilo/Monkey/monkey2/modules/" -I"/Users/danilo/Monkey/monkey2/modules/monkey/native" -I"/Users/danilo/Monkey/monkey2/tmp/" -DNDEBUG=1 -DBB_THREADS=1 -I"/Users/danilo/Monkey/monkey2/tmp/untitled3.buildv1.1.15/macos_release_mx/build/" -MMD -MF"/Users/danilo/Monkey/monkey2/tmp/untitled3.buildv1.1.15/macos_release_mx/build/untitled3bbec1512.cpp.deps" -o "/Users/danilo/Monkey/monkey2/tmp/untitled3.buildv1.1.15/macos_release_mx/build/untitled3bbec1512.cpp.o" "/Users/danilo/Monkey/monkey2/tmp/untitled3.buildv1.1.15/macos_release_mx/src/untitled3_untitled3.cpp"

In file included from /Users/danilo/Monkey/monkey2/tmp/untitled3.buildv1.1.15/macos_release_mx/src/untitled3_untitled3.cpp:2:
In file included from /Users/danilo/Monkey/monkey2/tmp/untitled3.buildv1.1.15/macos_release_mx/include/untitled3_untitled3.h:5:
In file included from /Users/danilo/Monkey/monkey2/modules/monkey/native/bbmonkey.h:14:
In file included from /Users/danilo/Monkey/monkey2/modules/monkey/native/bbvariant.h:5:
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:85:20: warning: binding dereferenced null pointer to reference has undefined behavior [-Wnull-dereference]
        return bbGetType( *(T*)0 );
                          ^~~~~~
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbvariant.h:69:11: note: in instantiation of function template specialization 'bbGetType<int>' requested here
                        return bbGetType<T>();
                               ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbvariant.h:61:3: note: in instantiation of member function 'bbVariant::Rep<int>::getType' requested here
                Rep( const T &value ):value( value ){
                ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbvariant.h:111:63: note: in instantiation of member function 'bbVariant::Rep<int>::Rep' requested here
        template<class T> explicit bbVariant( const T &t ):_rep( new Rep<T>( t ) ){
                                                                     ^
/Users/danilo/Monkey/monkey2/tmp/untitled3.buildv1.1.15/macos_release_mx/src/untitled3_untitled3.cpp:13:12: note: in instantiation of function template specialization 'bbVariant::bbVariant<int>' requested here
  bb_print(bbVariant(-2147483647).getType()->getName());
           ^
In file included from /Users/danilo/Monkey/monkey2/tmp/untitled3.buildv1.1.15/macos_release_mx/src/untitled3_untitled3.cpp:2:
In file included from /Users/danilo/Monkey/monkey2/tmp/untitled3.buildv1.1.15/macos_release_mx/include/untitled3_untitled3.h:5:
In file included from /Users/danilo/Monkey/monkey2/modules/monkey/native/bbmonkey.h:14:
In file included from /Users/danilo/Monkey/monkey2/modules/monkey/native/bbvariant.h:5:
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:85:9: error: call to 'bbGetType' is ambiguous
        return bbGetType( *(T*)0 );
               ^~~~~~~~~
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbvariant.h:69:11: note: in instantiation of function template specialization 'bbGetType<long>' requested here
                        return bbGetType<T>();
                               ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbvariant.h:61:3: note: in instantiation of member function 'bbVariant::Rep<long>::getType' requested here
                Rep( const T &value ):value( value ){
                ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbvariant.h:111:63: note: in instantiation of member function 'bbVariant::Rep<long>::Rep' requested here
        template<class T> explicit bbVariant( const T &t ):_rep( new Rep<T>( t ) ){
                                                                     ^
/Users/danilo/Monkey/monkey2/tmp/untitled3.buildv1.1.15/macos_release_mx/src/untitled3_untitled3.cpp:14:12: note: in instantiation of function template specialization 'bbVariant::bbVariant<long>' requested here
  bb_print(bbVariant(1l).getType()->getName());
           ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:66:1: note: candidate function
BB_GETTYPE_DECL( bbBool )
^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:64:45: note: expanded from macro 'BB_GETTYPE_DECL'
#define BB_GETTYPE_DECL( TYPE ) bbTypeInfo *bbGetType( TYPE const& );
                                            ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:67:1: note: candidate function
BB_GETTYPE_DECL( bbByte )
^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:64:45: note: expanded from macro 'BB_GETTYPE_DECL'
#define BB_GETTYPE_DECL( TYPE ) bbTypeInfo *bbGetType( TYPE const& );
                                            ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:68:1: note: candidate function
BB_GETTYPE_DECL( bbUByte )
^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:64:45: note: expanded from macro 'BB_GETTYPE_DECL'
#define BB_GETTYPE_DECL( TYPE ) bbTypeInfo *bbGetType( TYPE const& );
                                            ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:69:1: note: candidate function
BB_GETTYPE_DECL( bbShort )
^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:64:45: note: expanded from macro 'BB_GETTYPE_DECL'
#define BB_GETTYPE_DECL( TYPE ) bbTypeInfo *bbGetType( TYPE const& );
                                            ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:70:1: note: candidate function
BB_GETTYPE_DECL( bbUShort )
^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:64:45: note: expanded from macro 'BB_GETTYPE_DECL'
#define BB_GETTYPE_DECL( TYPE ) bbTypeInfo *bbGetType( TYPE const& );
                                            ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:71:1: note: candidate function
BB_GETTYPE_DECL( bbInt )
^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:64:45: note: expanded from macro 'BB_GETTYPE_DECL'
#define BB_GETTYPE_DECL( TYPE ) bbTypeInfo *bbGetType( TYPE const& );
                                            ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:72:1: note: candidate function
BB_GETTYPE_DECL( bbUInt )
^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:64:45: note: expanded from macro 'BB_GETTYPE_DECL'
#define BB_GETTYPE_DECL( TYPE ) bbTypeInfo *bbGetType( TYPE const& );
                                            ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:73:1: note: candidate function
BB_GETTYPE_DECL( bbLong )
^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:64:45: note: expanded from macro 'BB_GETTYPE_DECL'
#define BB_GETTYPE_DECL( TYPE ) bbTypeInfo *bbGetType( TYPE const& );
                                            ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:74:1: note: candidate function
BB_GETTYPE_DECL( bbULong )
^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:64:45: note: expanded from macro 'BB_GETTYPE_DECL'
#define BB_GETTYPE_DECL( TYPE ) bbTypeInfo *bbGetType( TYPE const& );
                                            ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:75:1: note: candidate function
BB_GETTYPE_DECL( bbFloat )
^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:64:45: note: expanded from macro 'BB_GETTYPE_DECL'
#define BB_GETTYPE_DECL( TYPE ) bbTypeInfo *bbGetType( TYPE const& );
                                            ^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:76:1: note: candidate function
BB_GETTYPE_DECL( bbDouble )
^
/Users/danilo/Monkey/monkey2/modules/monkey/native/bbtypeinfo.h:64:45: note: expanded from macro 'BB_GETTYPE_DECL'
#define BB_GETTYPE_DECL( TYPE ) bbTypeInfo *bbGetType( TYPE const& );
                                            ^
1 warning and 1 error generated.


***** Fatal mx2cc error *****

Internal mx2cc build error
```

**Comment by seyhajin** [comment]
Same error on Linux :(

**Comment by D-a-n-i-l-o** [comment]
Fixed with commit [f53dbc333db8527a2f2e7297daba95ae25654124](https://github.com/blitz-research/monkey2/commit/f53dbc333db8527a2f2e7297daba95ae25654124). Closing issue.
